File: LeetCode/0149. Max Points on a Line/solution.cpp
/**
 * 34 / 34 test cases passed.
 * Runtime: 16 ms
 * Memory Usage: 9.7 MB 
 */
class Solution {
public:
    int maxPoints(vector<vector<int>>& points) {
        int n = points.size();
        if (n <= 2) return n;
        int ans = 0;
        for (int i = 0; i < n - 1; ++ i) {
            if (ans > n / 2) return ans;
            
            unordered_map<string, int> cnt;
            int xi = points[i][0];
            int yi = points[i][1];
            for (int j = i + 1; j < n; ++ j) {
                int dx = points[j][0] - xi;
                int dy = points[j][1] - yi;
                if (dx == 0) {
                    dy = 1;
                } else if (dy == 0) {
                    dx = 1; 
                } else {
                    if (dy < 0) {
                        dx = -dx;
                        dy = -dy;
                    }
                    int g = __gcd(abs(dy), abs(dx));
                    dx /= g;
                    dy /= g;
                }
                string key = to_string(dy)+"/"+to_string(dx);
                cnt[key] += 1;
                ans = max(ans, cnt[key] + 1);
            }
        }
        return ans;
    }
};

File: LeetCode/0149. Max Points on a Line/solution.py
"""
34 / 34 test cases passed.
Runtime: 60 ms
Memory Usage: 14.9 MB
"""
class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        n, ans = len(points), 0
        if n <= 2: return n

        for i in range(n - 1):
            if ans > n // 2: return ans

            cnt = collections.defaultdict(int)
            xi, yi = points[i][0], points[i][1]
            for j in range(i + 1, n):
                dx, dy = points[j][0] - xi, points[j][1] - yi
                if dx == 0: dy = 1
                elif dy == 0: dx = 1
                else:
                    if dy < 0:
                        dx, dy = -dx, -dy
                    g = math.gcd(abs(dy), abs(dx))
                    dx //= g
                    dy //= g
                key = str(dy) + "/" + str(dx)
                cnt[key] += 1
                ans = max(ans, cnt[key] + 1)
        return ans
